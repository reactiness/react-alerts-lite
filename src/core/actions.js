export const actions = {
  add, 
  remove,
}

function add(alert) {
  return {
    type: 'ADD',
    payload: alert,
  }
};

function remove(alert) {
  return {
    type: 'REMOVE',
    payload: alert,
  }
};
