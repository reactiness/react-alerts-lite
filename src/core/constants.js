export const bottom = "bottom";
export const top = "top";
export const topLeft = "top-left";
export const topRight = "top-right";
export const bottomLeft = "bottom-left";
export const bottomRight = "bottom-right";
export const bottomFull = "bottom-full";
export const topFull = "top-full";

export const basic = "basic";
export const error = "error";
export const warning = "warning";
export const info = "info";
export const success = "success";

export const alertConstants = { basic, error, warning, info, success };

export const positionConstants = {
  bottom,
  top,
  topLeft,
  topRight,
  bottomLeft,
  bottomRight,
  bottomFull,
  topFull
};

export const positionsArray = [
  bottom,
  top,
  topLeft,
  topRight,
  bottomLeft,
  bottomRight,
  bottomFull,
  topFull
];

const fadeSlideDown = "fade-slide-down";
const fadeSlideUp = "fade-slide-up";
const fadeSlideDownUp = "fade-slide-down-up";
const fade = "fade";
const slideRight = "slide-right";
const slideLeft = "slide-left";
const rotateLeft = "rotate-left";
const rotateRight = "rotate-right";
const test = "test";

export const transitionConstants = {
  fadeSlideDown,
  fadeSlideUp,
  fadeSlideDownUp,
  fade,
  slideRight,
  slideLeft,
  rotateLeft,
  rotateRight,
  test
};

const content =
  "content: tent: define some con tent: define some con tent: define some con tent: define some condefine some content";

export const defaultAlertArgs = {
  type: info,
  timeout: 5000,
  transition: "fade-slide-up",
  duration: 500,
  position: bottom,
  content,
  closeButton: true,
  maxHeight: "200px"
};
