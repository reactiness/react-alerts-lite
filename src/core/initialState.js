export const initialState = {
    alerts: {
        current: []
    }
}