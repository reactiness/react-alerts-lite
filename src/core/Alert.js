import React, { Component } from "react";
import close from "../assets/x.svg";
import { CLASS_PREFIX } from "./constants";

const createHandlers = (remove, id) => {
  const handleRemove = () => {
    remove(id);
  };
  return {
    handleRemove
  };
};

export class Alert extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      renderChild: true
    };
    this.interval = null;
    this.handlers = createHandlers(props.remove, props.id);
  }

  componentDidMount() {
    this.interval = setTimeout(this.handlers.handleRemove, this.props.timeout);
  }

  componentWillUnmount() {
    const { onClose } = this.props;
    if (onClose) onClose();
    clearInterval(this.interval);
  }

  render() {
    const {
      content,
      type,
      style: transitionStyle,
      icon,
      closeButton,
      theme,
      className
    } = this.props;

    return (
      <div
        className={`${CLASS_PREFIX}container ${className}`}
        style={transitionStyle}
      >
        <div className={`${CLASS_PREFIX}wrapper ${theme} ${type}`}>
          <div className={`${CLASS_PREFIX}icon`}>
            <img src={icon} />
          </div>
          <div className={`${CLASS_PREFIX}content`}>{content}</div>
          {closeButton && (
            <div
              className={`${CLASS_PREFIX}close`}
              onClick={this.handlers.handleRemove}
            >
              <img src={close} />
            </div>
          )}
        </div>
      </div>
    );
  }
}
