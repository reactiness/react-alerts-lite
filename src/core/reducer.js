import {initialState} from './initialState';

const remove = (state, id) => {
  return {
    ...state,
    current: state.current.filter((el) => el.id === id)
  }
}

const add = (state, alert) => {
  return {
    ...state,
    current: [
      ...state.current,
      alert
    ]
  }
}

export function reducer(state = initialState.alerts, action) {
  switch (action.type) {
    case 'REMOVE': {
      return remove(state, action.payload)
    }

    case 'ADD': {
      return add(state, action.payload)
    }

    default:
      return state
  }
}