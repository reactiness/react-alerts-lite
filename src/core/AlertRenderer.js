import { Alert } from "./Alert";
import React, { Component } from "react";
import TransitionGroup from "react-transition-group/TransitionGroup";
import { FadeAndSlideTransition as DefaultTransition } from "../transitions/FadeSlide";
import { containerStyle, positionStyle } from "../css/styles";
import { positions } from "./constants";

export class AlertRenderer extends Component {
  render() {
    const { alerts, remove } = this.props;
    console.log("rendering alert renderer", alerts, positions);
    return positions.map(position => (
      <TransitionGroup
        style={{ ...positionStyle.common, ...positionStyle[position] }}
        key={`__rta_pos_${position}`}
      >
        {alerts[position].map((alert, i) => {
          const Transition = alert.transition
            ? alert.transition
            : DefaultTransition;
          return (
            <Transition
              duration={alert.duration}
              key={`__rta_alert_${alert.id}`}
            >
              <Alert pos={i} remove={remove} {...alert} />
            </Transition>
          );
        })}
      </TransitionGroup>
    ));
  }
}
