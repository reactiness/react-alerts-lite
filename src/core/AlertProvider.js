import React, { Component } from "react";
import { reducer } from "../state/reducer";
import { initialState } from "../state/initialState";
import { AlertRenderer } from "./AlertRenderer";
import { Store } from "../state/Store";
import { actions } from "../state/actions";
import { positionsArray } from "./constants";

const reducers = {
  alerts: reducer
};

export const store = new Store(reducers, initialState);

const createHandlers = dispatch => {
  const handleRemove = id => {
    dispatch(actions.remove(id));
  };
  const addDefaultProps = props => {
    dispatch(actions.addDefaultProps(props));
  };
  return {
    handleRemove,
    addDefaultProps
  };
};

const initEmptyPositionsObject = () => {
  const positionsObject = {};
  positionsArray.forEach(position => {
    positionsObject[position] = [];
  });
  return positionsObject;
};

export class AlertProvider extends Component {
  constructor(props) {
    super(props);
    this.state = {
      alerts: initEmptyPositionsObject()
    };
    this.handlers = createHandlers(store.dispatch);
  }

  sortAlerts(alerts, container) {
    alerts.forEach(alert => {
      container[alert.position].push(alert);
    });
    this.setState({ alerts: container });
  }

  componentDidMount() {
    store.subscribe(store =>
      this.sortAlerts(store.alerts.current, initEmptyPositionsObject())
    );
    if (this.props.defaultProps)
      this.handlers.addDefaultProps(this.props.defaultProps);
  }

  render() {
    return (
      <AlertRenderer
        alerts={this.state.alerts}
        remove={this.handlers.handleRemove}
        transitions={this.props.transitions}
      />
    );
  }
}
