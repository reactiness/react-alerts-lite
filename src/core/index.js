import React from "react";
import { Alert } from "./Alert";
import { store } from "./AlertProvider";
import { actions } from "../state/actions";
import { defaultAlertArgs } from "./constants";

let counter = 0;

const add = args => {
  console.log(store);
  store.dispatch(
    actions.add({
      id: counter,
      ...args
    })
  );
  counter++;
};

const remove = type => {
  type
    ? store.dispatch(actions.removeType(type))
    : store.dispatch(actions.removeAll());
};

export class Alerts {
  static push(args) {
    console.log();
    add({ ...defaultAlertArgs, ...store.state.alerts.defaultProps, ...args });
  }
  static remove(type) {
    remove(type);
  }
}
