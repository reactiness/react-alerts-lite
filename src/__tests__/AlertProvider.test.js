import React from "react";
import { AlertProvider } from "../core/AlertProvider.js";
import { positionConstants } from "../core/constants";

const alert = {
  id: 0,
  closeButton: true,
  content: "this is a test alert",
  duration: 500,
  maxHeight: "200px",
  position: "top-right",
  timeout: 3000,
  transition: "scale-slide-right",
  type: "basic"
};

const populate = () => {
  return [
    alert,
    { ...alert, id: 1, type: "basic", position: positionConstants.topLeft },
    { ...alert, id: 2, type: "basic", position: positionConstants.topRight },
    { ...alert, id: 3, type: "basic", position: positionConstants.bottomLeft },
    { ...alert, id: 4, type: "basic", position: positionConstants.bottomRight },
    { ...alert, id: 5, type: "basic", position: positionConstants.topFull },
    { ...alert, id: 6, type: "basic", position: positionConstants.bottomFull },
    { ...alert, id: 7, type: "basic", position: positionConstants.top },
    { ...alert, id: 8, type: "basic", position: positionConstants.bottom },
    { ...alert, id: 9, type: "basic", position: positionConstants.top }
  ];
};

const manuallySorted = {
  bottom: [
    { ...alert, id: 8, type: "basic", position: positionConstants.bottom }
  ],
  top: [
    { ...alert, id: 7, type: "basic", position: positionConstants.top },
    { ...alert, id: 9, type: "basic", position: positionConstants.top }
  ],
  "top-left": [
    { ...alert, id: 1, type: "basic", position: positionConstants.topLeft }
  ],
  "top-right": [
    alert,
    { ...alert, id: 2, type: "basic", position: positionConstants.topRight }
  ],
  "bottom-left": [
    { ...alert, id: 3, type: "basic", position: positionConstants.bottomLeft }
  ],
  "bottom-right": [
    { ...alert, id: 4, type: "basic", position: positionConstants.bottomRight }
  ],
  "bottom-full": [
    { ...alert, id: 6, type: "basic", position: positionConstants.bottomFull }
  ],
  "top-full": [
    { ...alert, id: 5, type: "basic", position: positionConstants.topFull }
  ]
};

describe("Alert Provider", () => {
  it("should render correctly", () => {
    const wrapper = shallow(<AlertProvider />);
    expect(wrapper).toMatchSnapshot();
  });
  it("should sort alerts correctly", () => {
    const wrapper = mount(<AlertProvider />);
    const alerts = populate();
    const positions = wrapper.instance().initEmptyPositionsObject();
    const sorted = wrapper.instance().sortAlerts(alerts, positions);
    expect(wrapper.state("alerts")).toEqual(manuallySorted);
  });
  // it("should add default props if they are passed in as a prop", () => {
  //   const defaultProps = {
  //     transition: "fade",
  //     duration: 500
  //   };
  //   const wrapper = mount(<AlertProvider defaultProps={defaultProps} />);
  //   console.log('proto', wrapper.instance().handlers);
  //   const mock = jest.spyOn(wrapper.instance(), "handleAddDefaultProps");
  //   // console.log('INSTNCE', wrapper.instance());
  //   expect(mock).toHaveBeenCalledTimes(1);
  // });
});
