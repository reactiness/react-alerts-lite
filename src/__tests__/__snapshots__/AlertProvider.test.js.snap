// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Provider should render correctly 1`] = `
<AlertRenderer
  alerts={
    Object {
      "bottom": Array [],
      "bottom-full": Array [],
      "bottom-left": Array [],
      "bottom-right": Array [],
      "top": Array [],
      "top-full": Array [],
      "top-left": Array [],
      "top-right": Array [],
    }
  }
  remove={[Function]}
/>
`;
